### PR TITLE
Dev/saars/k8s client

### PR DESCRIPTION
**Pull request by xiaomi7732**
Adress #288. Please check out the tested scenarios in it.

Also, the target branch expects to be feature/3.0.

### CODE REVIEW

File: src/ApplicationInsights.Kubernetes/EntityExtensions/PodExtensions.cs [hunk 1]
```diff
@@ -1,49 +1,43 @@
...
-            {
-                K8sReplicaSet? replica = scope?.FirstOrDefault(
-                    r => r.Metadata != null &&
-                    r.Metadata.Uid != null &&
-                    r.Metadata.Uid.Equals(replicaRef.Uid, StringComparison.OrdinalIgnoreCase));
-                return replica;
-            }
-            return null;
+            V1ReplicaSet? replica = scope?.FirstOrDefault(
+                r => string.Equals(r.Metadata.Uid, replicaRef.Uid, StringComparison.OrdinalIgnoreCase));
```
Comment (xiaomi7732) [inline src/ApplicationInsights.Kubernetes/EntityExtensions/PodExtensions.cs:38]: It is interesting to find out, Uid is of type "GUID", and guid is case insensitive. :-)

File: src/ApplicationInsights.Kubernetes/EntityExtensions/PodExtensions.cs [hunk 2]
```diff
@@ -1,49 +1,43 @@
...
+﻿using System;
+using System.Linq;
+using System.Collections.Generic;
+using k8s.Models;
 
-namespace Microsoft.ApplicationInsights.Kubernetes
+namespace Microsoft.ApplicationInsights.Kubernetes;
+
+// Extension methods for Pod
+internal static class PodExtensions
```
Comment (xiaomi7732) [inline src/ApplicationInsights.Kubernetes/EntityExtensions/PodExtensions.cs:9]: Consider: should this be part of `IPodInfoManager`?

File: src/ApplicationInsights.Kubernetes/IK8sQueryClient.cs [hunk 3]
```diff
@@ -1,19 +1,16 @@
...
+﻿using System.Collections.Generic;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.ApplicationInsights.Kubernetes.Entities;
+using k8s.Models;
 
 namespace Microsoft.ApplicationInsights.Kubernetes
 {
-    internal interface IK8sQueryClient : IDisposable
+    internal interface IK8sQueryClient
```
Comment (xiaomi7732) [inline src/ApplicationInsights.Kubernetes/IK8sQueryClient.cs:8]: How much value is this interface provide? Is this the same as IK8sClientService? Why do we need this in the middle?

File: src/ApplicationInsights.Kubernetes/TelemetryInitializers/TelemetryKeyCache.cs [hunk 4]
```diff
@@ -8,26 +8,27 @@ namespace Microsoft.ApplicationInsights.Kubernetes
     internal class TelemetryKeyCache : ITelemetryKeyCache
     {
         private readonly AppInsightsForKubernetesOptions _options;
-
-        private ConcurrentDictionary<string, string> _telemetryKeyCache;
+        private readonly ConcurrentDictionary<string, string> _telemetryKeyCache = new ConcurrentDictionary<string, string>();
```
Comment (xiaomi7732) [inline src/ApplicationInsights.Kubernetes/TelemetryInitializers/TelemetryKeyCache.cs:11]: Why initialize here and then again in the ctor? It looks like it is also conflict with the null check in GetProcessedKey method. Should this be a nullable concurrent dictionary?

**Review by xiaomi7732 — COMMENTED** [review]
Leave some questions to consider.

File: dev/F5WebApi/Readme.md [hunk 5]
```diff
@@ -48,3 +48,18 @@
     ```shell
     kubectl create -f ./k8s.yml
     ```
+
+1. Port forward for pod for local testing:
```
Comment (xiaomi7732) [inline dev/F5WebApi/Readme.md:52]: Port forward to the pod for local testing:

**Review by karolz-ms — APPROVED** [review]
How simpler this is! 😍

File: src/ApplicationInsights.Kubernetes/Containers/IContainerStatusManager.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+using System.Threading;
+using System.Threading.Tasks;
+using k8s.Models;
+
+namespace Microsoft.ApplicationInsights.Kubernetes.Containers;
+
+internal interface IContainerStatusManager
+{
+    Task<V1ContainerStatus?> TryGetMyContainerStatusAsync(CancellationToken cancellationToken);
```
[the remark this answers is not in the document]
Comment (xiaomi7732) [reply src/ApplicationInsights.Kubernetes/Containers/IContainerStatusManager.cs:9]: That makes sense.

[the remark this answers is not in the document]
Comment (xiaomi7732) [reply src/ApplicationInsights.Kubernetes/Containers/IContainerStatusManager.cs]: Fixed.

**Pull request merged**